brainscore/metrics/assembly_utils.py

```diff
@@ -0,0 +1,251 @@
+                xarr.append(arr)
+       
+        if not xarr:
+            return
+        else:
+            xarr = xr.concat(xarr, dim=dim)
+            attrs = {k: xr.concat(vs, dim=dim) for k,vs in attrs.items()}
+            xarr.coords[dim] = asms[0].coords[dim]
+            for k,v in attrs.items():
+                attrs[k].coords[dim] = asms[0].coords[dim]
+                xarr.attrs[k] = attrs[k]
+            return xarr
+
+    return apply_helper(sizes, dims, *asms)
+
+
+
+if __name__ == "__main__":
+
+    ## align
+
+    x = DataAssembly(
+        [
+            [1, 2],
+            [3, 4],
+        ],
+        dims=["a", "b"],
+        coords={"a_0": ("a", [0, 1]), "a_1": ("a", [0, 1]), "b": [0, 1]}
+    )
+
+    y = DataAssembly(
+        [
+            [5, 7],
+            [6, 8]
+        ],
+        dims=["b", "a"],
+        coords={"a": [0, 2], "b": [0, 1]}
+    ) 
+
+    z = DataAssembly(
+        [
+            [[2], [4], [99]],
+            [[1], [3], [100]],
+        ],
+        dims=["b", "a", "c"],
+        coords={"a_0": ("a", [0, 1, 2]), "a_1": ("a", [0, 1, 2]), "b": [1, 0], "c": [10]}
+    )
+
+    d = DataAssembly(
+        [
+            [[2], [4]],
+            [[1], [3]],
+        ],
+        dims=["d", "a", "c"],
+        coords={"a_0": ("a", [0, 1]), "a_1": ("a", [0, 1]), "d": [1, 0], "c": [10]}
+    )
+
+    def _except(func, *args, **kwargs):
+        try:
+            func(*args, **kwargs)
+        except Exception as e:
+            print(e)
+        else:
+            assert False
+
+    _except(align_dims, x, y, dims=["a", "b"])
+    x_, y_ = align_dims(x, y, dims=["b"])
+    assert (x.transpose("b", ...).sortby(["b"]) == x_).all()
+    assert (y==y_).all()
+    x_, z_ = align_dims(x, z, dims=["a", "b"])
+    assert (x.sortby(["a", "b"]) == x_).all()
+    assert (x.sortby(["a", "b"]).values == z_.isel(c=0).values).all()
+    _except(align_dims, x, d, dims=["a", "b"])
+    _except(align_dims, x, z, dims=["a", "b"], dims_must_equal=True)
+
+    ## compute_over_dim
+
+    from functools import partial
+    check_coords_equal = partial(check_coords_contain, equal=True)
+
+    assert check_coords_equal(x, x.transpose("b", "a"))
+    assert not check_coords_equal(x, y)
+    assert not check_coords_equal(x, z.sortby("b"))
+
+    import numpy as np
+    arr = DataAssembly([[1, 2], [3, 4]], dims=["x", "y"], coords={"x_ab": ("x", ["a", "b"]), "x_01": ("x", [0, 1]), "y": [0, 1]})
+    def func(ele):
+        ret = ele ** 2
+        ret.attrs["origin"] = ele
+        return ret
+
+    ret = apply_over_dims(func, arr, dims=["x", "y"])
+    assert (ret.values == np.array([[1, 4], [9, 16]])).all()
+    assert (ret.attrs["origin"].values == arr.values).all()
+    assert check_coords_equal(ret, arr)
+
+    ret = apply_over_dims(func, arr, dims=["x"])
+    assert (ret.values == np.array([[1, 4], [9, 16]])).all()
+    assert (ret.attrs["origin"].values == arr.values).all()
+    assert check_coords_equal(ret, arr)
```

Comment:
these tests are great, but let's put them inside unit tests in the `tests/` directory